### ISSUE TITLE
Using CMake CUDA language downstream

### ISSUE THREAD

**Issue by ibaned** [kokkos/kokkos#4319]
In our projects, we use the CMake support for CUDA to automatically decide how to compile our files, without the need for wrappers like `nvcc_wrapper` and `kokkos_launch_compiler`.

In order to use Kokkos, we would like to have an option to do `find_package(Kokkos)` in CMake _without_ triggering the call to `kokkos_compilation` that sets `RULE_LAUNCH_COMPILE` at a global level or on my targets.

Maybe this is already possible but we would like to know an officially supported and reliable way to not have `RULE_LAUNCH_COMPILE` etc affected in our downstream CMake files.

**Comment by ibaned** [comment]
So I found the `separable_compilation` component thing in the logic, and did:
```cmake
find_package(Kokkos COMPONENTS separable_compilation)
```
This seems to allow me to avoid `kokkos_compilation` and `kokkos_launch_compiler`, however:

In CMake, CUDA only affects the compilation of source files into objects and not the linking of binary objects together. This means that an executable is linked by the C++ compiler, and so the flags on the `Kokkos::kokkos` target become a problem:

```
[ 83%] Building CUDA object CMakeFiles/p3a-unit-tests.dir/p3a_unit_tests.cpp.o
[100%] Linking CXX executable p3a-unit-tests
g++: error: unrecognized command line option ‘-arch=sm_70’
```

**Comment by crtrott** [comment]
You could probably set CMAKE_LINKER=nvcc? There are some cases where you actually need to do so (for example with RDC on). But I think the real issue is that we need for your purpose to attache flags to the CMAKE_CUDA_FLAGS not the CMAKE_CXX_FLAGS which are actually not used for CUDA object files.

**Comment by ibaned** [comment]
In this case it comes from the following on the `Kokkos::kokkos` target:
```cmake
INTERFACE_LINK_OPTIONS "\$<\$<LINK_LANGUAGE:CXX>:-DKOKKOS_DEPENDENCE;-arch=sm_70>"
```
Is this option actually needed if the linker is `nvcc` and if so can we introduce it through the `RULE_LAUNCH_LINK` thing in `kokkos_compilation` instead of the `INTERFACE_LINK_OPTIONS`?

**Comment by masterleinad** [comment]
We added the architecture to the link options to avoid some `nvcc` linker warnings regarding deprecated architectures. I am not quite sure if we can use `RULE_LAUNCH_LINK` for this sind we would want the same behavior when using `nvcc_wrapper` directly.

**Comment by ibaned** [comment]
Can we come up with an option to build Kokkos such that it only supports either the `kokkos_compilation` macro or the CMake CUDA language but does not support using `nvcc_wrapper` directly? If a single build doesn't have to support all the different downstream approaches then it can make better choices

**Comment by ibaned** [comment]
Another option is to provide another target. I'm happy to link to `Kokkos::kokkos_modern_cmake` which doesn't have the CUDA flags on it

**Comment by crtrott** [comment]
Btw. how does this work with Clang as CUDA compiler?

**Comment by crtrott** [comment]
regarding the linker option we can just add an umbrella option which turns that off and whatever else we need to change.

**Comment by crtrott** [comment]
I still think we need in that case to add all our flags to the CMAKE_CUDA_FLAGS only

**Comment by masterleinad** [comment]
> Btw. how does this work with Clang as CUDA compiler?

We don't add the architecture flag to the link flags in that case.

**Comment by crtrott** [comment]
I think I have this almost figured out.

**Comment by crtrott** [comment]
Yeah I basically got it - minus the cleanup.

**Comment by crtrott** [comment]
Got a PR ^^

**Comment by PhilMiller** [comment]
Fix released in 3.6
